includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -349,230 +431,741 @@ private function process_link_element( DOMElement $element ) {
+			foreach ( $parsed['validation_errors'] as $validation_error ) {
+				call_user_func( $this->args['validation_error_callback'], array_merge( $validation_error, compact( 'node' ) ) );
+			}
+		}
+
+		$stylesheet = '';
+		foreach ( $parsed['stylesheet'] as $stylesheet_part ) {
+			if ( is_array( $stylesheet_part ) ) {
+				list( $selectors_parsed, $declaration_block ) = $stylesheet_part;
+				if ( $should_tree_shake ) {
+					$selectors = array();
+					foreach ( $selectors_parsed as $selector => $class_names ) {
+						if ( 0 === count( array_diff( $class_names, $this->used_class_names ) ) ) { // If all class names are used in the doc.
+							$selectors[] = $selector;
+						}
+					}
+				} else {
+					$selectors = array_keys( $selectors_parsed );
+				}
+				if ( ! empty( $selectors ) ) {
+					$stylesheet .= implode( ',', $selectors ) . $declaration_block;
+				}
+			} else {
+				$stylesheet .= $stylesheet_part;
+			}
 		}
+
 		return $stylesheet;
 	}
 
 	/**
-	 * Validate amp-keyframe style.
+	 * Parse stylesheet.
 	 *
-	 * @since 0.7
-	 * @link https://github.com/ampproject/amphtml/blob/b685a0780a7f59313666225478b2b79b463bcd0b/validator/validator-main.protoascii#L1002-L1043
+	 * @since 1.0
+	 *
+	 * @param string $stylesheet_string Stylesheet.
+	 * @param array  $options           Options. See definition in \AMP_Style_Sanitizer::process_stylesheet().
+	 * @return array {
+	 *    Parsed stylesheet.
 	 *
-	 * @param DOMElement $style Style element.
-	 * @return true|WP_Error Validity.
+	 *    @type array $stylesheet        Stylesheet parts, where arrays are tuples for declaration blocks.
+	 *    @type array $validation_errors Validation errors.
+	 * }
 	 */
-	private function validate_amp_keyframe( $style ) {
-		if ( 'body' !== $style->parentNode->nodeName ) {
-			return new WP_Error( 'mandatory_body_child', __( 'amp-keyframes is not child of body element.', 'amp' ) );
-		}
+	private function parse_stylesheet( $stylesheet_string, $options = array() ) {
+		$start_time = microtime( true );
+
+		$options = array_merge(
+			array(
+				'allowed_at_rules'           => array(),
+				'convert_width_to_max_width' => false,
+				'property_blacklist'         => array(
+					// See <https://www.ampproject.org/docs/design/responsive/style_pages#disallowed-styles>.
+					'behavior',
+					'-moz-binding',
+				),
+				'property_whitelist'         => array(),
+				'validate_keyframes'         => false,
+				'stylesheet_url'             => null,
+				'stylesheet_path'            => null,
+			),
+			$options
+		);
 
-		if ( $this->keyframes_max_size && strlen( $style->textContent ) > $this->keyframes_max_size ) {
-			return new WP_Error( 'max_bytes', __( 'amp-keyframes is too large', 'amp' ) );
-		}
+		$stylesheet        = array();
+		$validation_errors = array();
+		try {
+			$parser_settings = Sabberworm\CSS\Settings::create()->withMultibyteSupport( false );
```

Comment:
I think `->withMultibyteSupport( false )` can be removed. I only added it to see if it would improve performance, but it only helped marginally, and since the results are cached it doesn't matter so much.